Add optional parameter to like(), or_like(), not_like() and or_not_like() methods.

nalakapws opened the pull request:
This PR address the 2nd feature request from this #5641 issue.

nalakapws committed:
Add optional parameter called $clause to like(), or_like(), not_like() and or_not_like() methods. which can be used to choose whether use those methods with WHERE clause or HAVING clause.

narfbg commented:
This is parameter creep ... It may look inconsistent, but a better option would be to just have `having_like()`, `having_not_like()`, etc.

And you must follow the [styleguide](https://www.codeigniter.com/user_guide/general/styleguide.html).

nalakapws committed:
Introduce new having_like(), or_having_like(), not_having_like() and or_not_having_like() methods.

nalakapws committed:
Trim trailing whitespaces and convert indentation to spaces(4).

nalakapws committed:
Cleanup whitespaces(Spaces and Tabs) according to Codeigniter's style guide.

Nudasoft commented:
Is there anything I have to do in order to make this PR get merged?

narfbg commented:
Yes, there are still some spaces left in on lines 1100, 1171 (shown in diff with darker green).

And there's more significant stuff code-wise, but it will probably result in a lot of back and forth if I asked you to do all of it, so I can make those changes myself after merging. It's decent enough as it is, I just want less repetition.

nalakapws committed:
Introduce new having_group_start(), or_having_group_start(), not_having_group_start(), or_not_having_group_start() and having_group_end() methods for HAVING clause.

NallDev commented on system/database/DB_query_builder.php:
```diff
@@ -999,24 +999,105 @@ public function or_not_like($field, $match = '', $side = 'both', $escape = NULL)
+	 * @param	bool	$escape
+	 * @return	CI_DB_query_builder
+	 */
+	public function having_like($field, $match = '', $side = 'both', $escape = NULL)
```
this line doesn't match PSR-2 rules

narfbg replied:
We don't subscribe to PSR-2.

NallDev commented on system/database/DB_query_builder.php:
```diff
@@ -999,24 +999,105 @@ public function or_not_like($field, $match = '', $side = 'both', $escape = NULL)
+		return $this->_like($field, $match, 'AND ', $side, '', $escape, 'having');
+	}
+
+	// --------------------------------------------------------------------
```
bad comment

narfbg replied:
It's not a "bad comment"; we have these in between all functions.

NallDev commented on system/database/DB_query_builder.php:
```diff
@@ -999,24 +999,105 @@ public function or_not_like($field, $match = '', $side = 'both', $escape = NULL)
+	 * @param	bool	$escape
+	 * @return	CI_DB_query_builder
+	 */
+	public function not_having_like($field, $match = '', $side = 'both', $escape = NULL)
```
this line doesn't match PSR-2 rules

narfbg replied:
This should be `having_not_like` instead.
Same thing applies to all other methods below - `not` should always be after `having` in the name.

NallDev commented on system/database/DB_query_builder.php:
```diff
@@ -999,24 +999,105 @@ public function or_not_like($field, $match = '', $side = 'both', $escape = NULL)
+	 * @param	bool	$escape
+	 * @return	CI_DB_query_builder
+	 */
+	public function or_having_like($field, $match = '', $side = 'both', $escape = NULL)
```
this line doesn't match PSR-2 rules

NallDev commented on system/database/DB_query_builder.php:
```diff
@@ -999,24 +999,105 @@ public function or_not_like($field, $match = '', $side = 'both', $escape = NULL)
+	 * @param	bool	$escape
+	 * @return	CI_DB_query_builder
+	 */
+	public function or_not_having_like($field, $match = '', $side = 'both', $escape = NULL)
```
this line doesn't match PSR-2 rules

NallDev commented on system/database/DB_query_builder.php:
```diff
@@ -1164,6 +1263,97 @@ public function group_end()
+	 * @param	string	$type	(Internal use only)
+	 * @return	CI_DB_query_builder
+	 */
+	public function having_group_start($not = '', $type = 'AND ')
```
this line doesn't match PSR-2 rules

NallDev commented on system/database/DB_query_builder.php:
```diff
@@ -1164,6 +1263,97 @@ public function group_end()
+
+		$this->qb_where_group_started = TRUE;
+		$prefix = (count($this->qb_having) === 0 && count($this->qb_cache_having) === 0) ? '' : $type;
+		$having = array(
```
You can use `[]`

narfbg replied:
... and I'd force him to change it back to `array()`, because that's what we use.

error56 reviewed (changes requested):
All the code doesn't match PSR-2 rules

narfbg commented:
Yes, all of CodeIgniter doesn't match PSR-2 rules. We have our own [styleguide](https://www.codeigniter.com/user_guide/general/styleguide.html), referenced in the [contributing guide](https://github.com/bcit-ci/CodeIgniter/blob/develop/contributing.md), which you should read before reviewing others' work.
